### ISSUE TITLE
WeakMapConstructor and WeakSetConstructor should be able to take `undefined` and `null`

### ISSUE THREAD

**Issue by AlexErrant** [microsoft/TypeScript#54641]
# lib Update Request

## Configuration Check

My compilation *target* is `ES2015` and my *lib* is `es2021`.

## Missing / Incorrect Definition

The `es2015.iterable` type definitions for `WeakMap` and `WeakSet`'s ctors are missing `undefined` and `null`, contrary to [the weakmap spec](https://tc39.es/ecma262/#sec-weakmap-iterable) and [the weakset spec](https://tc39.es/ecma262/#sec-weakset-iterable):

> 4. If iterable is either undefined or null, return map.
> 4. If iterable is either undefined or null, return set.

https://github.com/microsoft/TypeScript/blob/03951f2ddf727f6e697972a4e9589803229564b7/src/lib/es2015.iterable.d.ts#L146-L148

https://github.com/microsoft/TypeScript/blob/03951f2ddf727f6e697972a4e9589803229564b7/src/lib/es2015.iterable.d.ts#L194-L196

The (non-weak) Map ctor was made optional/nullable in [this PR](https://github.com/microsoft/TypeScript/pull/43396) from 2021 - I assume that `WeakMap`/`WeakSet` was missed as an oversight.

I began to address this issue in [this branch](https://github.com/microsoft/TypeScript/compare/main...AlexErrant:weakmap-weakset-nullable-ctor?expand=1), but the PR template required an issue so here I am :)

**Comment by RyanCavanaugh** [comment]
What are you trying to address? This is already legal:
```ts
// Not an error
const p1 = new WeakMap(null);
// Not an error
const p2 = new WeakMap(undefined);
// Not an error
const p3 = new WeakSet(null);
// Not an error
const p4 = new WeakSet(undefined);
```

**Comment by fatcerberus** [comment]
That’s interesting, since the only construct signature defined (as shown in the quoted type definition) is one that takes an `Iterable`

**Comment by AlexErrant** [comment]
I ran into this issue in the course of [this PR](https://github.com/solidjs-community/solid-primitives/pull/459). I didn't run into this issue in the course of run/compile-time analysis; I only, uh, noticed it during "copy/paste time analysis". I noticed it while trying to copy TypeScript's Map/Set constructor type definitions to the above library's Map/Set wrapper. (The above library defines a special Map/Set wrapper that something special in the SolidJS UI framework.)

I believe your example works because it uses `WeakMapConstructor`and `WeakSetConstructor` from https://github.com/microsoft/TypeScript/blob/main/src/lib/es2015.collection.d.ts , which correctly handles null/undefined. My issue is with `es2015.iterable`.
